src/dbnode/storage/index.go

```diff
@@ -501,31 +513,60 @@ func (i *nsIndex) writeBatches(
 		// on the provided inserts to terminate quicker during shutdown.
 		return
 	}
-	now := i.nowFn()
-	futureLimit := now.Add(1 * i.bufferFuture)
-	pastLimit := now.Add(-1 * i.bufferPast)
+	var (
+		now                 = i.nowFn()
+		blockSize           = i.blockSize
+		futureLimit         = now.Add(1 * i.bufferFuture)
+		pastLimit           = now.Add(-1 * i.bufferPast)
+		batchOptions        = batch.Options()
+		forwardIndexDice    = i.forwardIndexDice
+		forwardIndexEnabled = forwardIndexDice.enabled
+
+		forwardIndexBatch *index.WriteBatch
+	)
 	// NB(r): Release lock early to avoid writing batches impacting ticking
 	// speed, etc.
 	// Sometimes foreground compaction can take a long time during heavy inserts.
 	// Each lookup to ensureBlockPresent checks that index is still open, etc.
 	i.state.RUnlock()
 
+	if forwardIndexEnabled {
+		// NB(arnikola): Don't initialize forward index batch if forward indexing
+		// is not enabled.
+		forwardIndexBatch = index.NewWriteBatch(batchOptions)
+	}
 	// Ensure timestamp is not too old/new based on retention policies and that
-	// doc is valid.
+	// doc is valid. Add potential forward writes to the forwardWriteBatch.
 	batch.ForEach(
 		func(idx int, entry index.WriteBatchEntry,
 			d doc.Document, _ index.WriteBatchEntryResult) {
-			if !futureLimit.After(entry.Timestamp) {
+			ts := entry.Timestamp
+			if !futureLimit.After(ts) {
 				batch.MarkUnmarkedEntryError(m3dberrors.ErrTooFuture, idx)
 				return
 			}
 
-			if !entry.Timestamp.After(pastLimit) {
+			if !ts.After(pastLimit) {
 				batch.MarkUnmarkedEntryError(m3dberrors.ErrTooPast, idx)
 				return
 			}
+
+			if forwardIndexDice.roll(ts) {
```

Comment:
should this also be `if forwardIndexEnabled && ..` 

maybe add a test that would have caught this

Reply:
The `!enabled` case should be caught here: https://github.com/m3db/m3/pull/1613/files/70b0912726891694461e47b7d21499aeddb8bbca#diff-b5380446b59ebd51b8b8a482df15c4d7R93